pkg/fqdn/cache.go

```diff
@@ -154,16 +181,124 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
+// runs to the give time. It will lock the struct until retrieves all the data.
+// It returns the list of names that need to be deleted from the policies.
+func (c *DNSCache) cleanupExpiredEntries(expires time.Time) ([]string, time.Time) {
+	timediff := int(expires.Sub(c.lastCleanup).Seconds())
+	startTime := c.lastCleanup
+	expiredEntries := []string{}
+	for i := 0; i < int(timediff); i++ {
+		expiredTime := c.lastCleanup.Add(time.Second)
+		key := expiredTime.Unix()
+		c.lastCleanup = expiredTime
+		entries, exists := c.cleanup[key]
+		if !exists {
+			continue
+		}
+		expiredEntries = append(expiredEntries, entries...)
+		delete(c.cleanup, key)
+	}
+
+	result := []string{}
+	for _, name := range KeepUniqueNames(expiredEntries) {
+		if entries, exists := c.forward[name]; exists {
+			if !c.removeExpired(entries, c.lastCleanup, time.Time{}) {
+				continue
+			}
+			result = append(result, name)
+		}
+	}
+	return result, startTime
+}
+
+// cleanupOverLimitEntries returns the names that has reached the max number of
+// IP per host.
+// Internally the function sort the entries by the TTL.
+func (c *DNSCache) cleanupOverLimitEntries() []string {
+	type IPEntry struct {
+		ip    string
+		entry *cacheEntry
+	}
+	toClean := []string{}
+
+	// For global cache the limit maybe is not used at all.
+	if c.limit == 0 {
+		return toClean
+	}
+
+	for dnsName := range c.overLimit {
+		entries, ok := c.forward[dnsName]
+		if !ok {
+			continue
+		}
+		overlimit := len(entries) - FQDNMaxIPperHost
+		if overlimit <= 0 {
+			continue
+		}
+
+		sortedEntries := []IPEntry{}
+		for ip, entry := range entries {
+			sortedEntries = append(sortedEntries, IPEntry{ip, entry})
+		}
+
+		sort.Slice(sortedEntries, func(i, j int) bool {
+			return sortedEntries[i].entry.ExpirationTime.Unix() < sortedEntries[j].entry.ExpirationTime.Unix()
+		})
+
+		for i := 0; i <= overlimit; i++ {
+			key := sortedEntries[i]
+			delete(entries, key.ip)
+			c.removeReverse(key.ip, key.entry)
+		}
+		toClean = append(toClean, dnsName)
+	}
+	c.overLimit = map[string]bool{}
+	return toClean
+}
+
+// GC garbage collector function that clean expired entries and return the
+// entries that are deleted.
+func (c *DNSCache) GC() []string {
+	c.Lock()
+	expiredEntries, _ := c.cleanupExpiredEntries(time.Now())
+	overLimitEntries := c.cleanupOverLimitEntries()
+	c.Unlock()
+	return KeepUniqueNames(append(expiredEntries, overLimitEntries...))
+}
+
 // UpdateFromCache is a utility function that allows updating a DNSCache
 // instance with all the internal entries of another. Latest-Expiration still
 // applies, thus the merged outcome is consistent with adding the entries
-// individually.
-func (c *DNSCache) UpdateFromCache(update *DNSCache) {
+// individually. If any item in the names slice will only update that names.
+// If the clean argument is set to true, all entries will be clean.
+func (c *DNSCache) UpdateFromCache(update *DNSCache, names []string, clean bool) {
```

Comment:
names might be better as namesToUpdate or something,

the comment is a little unclear, maybe:
When names is non-nil only those names are updated from update, otherwise all DNS names in update are used.
When clean is true the data in c for each specific name is cleared before the update.